substrate/frame/asset-conversion/src/lib.rs

```diff
@@ -765,22 +772,125 @@ pub mod pallet {
+				ensure!(!amount_in.is_zero(), Error::<T>::ZeroAmount);
+				ensure!(amount_out_min.map_or(true, |a| !a.is_zero()), Error::<T>::ZeroAmount);
+
+				Self::validate_swap_path(&path)?;
+				let path = Self::balance_path_from_amount_in(amount_in, path)?;
+
+				let amount_out = path.last().map(|(_, a)| *a).ok_or(Error::<T>::InvalidPath)?;
+				ensure!(
+					amount_out_min.map_or(true, |a| amount_out >= a),
+					Error::<T>::ProvidedMinimumNotSufficientForSwap
+				);
+				Ok((path, amount_out))
+			};
+			let (path, amount_out) = match inspect_path(credit_in.asset()) {
+				Ok((p, a)) => (p, a),
+				Err(e) => return Err((credit_in, e)),
+			};
+
+			let credit_out = Self::credit_swap(credit_in, &path)?;
+
+			Self::deposit_event(Event::CreditSwapExecuted { amount_in, amount_out, path });
+
+			Ok(credit_out)
+		}
+
+		/// Swaps a portion of `credit_in` of `path[0]` asset to obtain the desired `amount_out` of
+		/// the `path[last]` asset. The provided `credit_in` must be adequate to achieve the target
+		/// `amount_out`, or an error will occur.
+		///
+		/// On success, the function returns a (`credit_out`, `credit_change`) tuple, where
+		/// `credit_out` represents the acquired amount of the `path[last]` asset, and
+		/// `credit_change` is the remaining portion from the `credit_in`. On failure, an `Err` with
+		/// the initial `credit_in` and error code is returned.
+		///
+		/// WARNING: This may return an error after a partial storage mutation. It should be used
+		/// only inside a transactional storage context and an Err result must imply a storage
+		/// rollback.
+		pub(crate) fn do_swap_credit_tokens_for_exact_tokens(
+			path: BoundedVec<T::MultiAssetId, T::MaxSwapPathLength>,
+			credit_in: Credit<T>,
+			amount_out: T::AssetBalance,
+		) -> Result<(Credit<T>, Credit<T>), (Credit<T>, DispatchError)> {
+			let amount_in_max = match credit_in.peek() {
+				Ok(a) => a,
+				Err(e) => return Err((credit_in, e)),
+			};
+			let inspect_path = |credit_asset| {
+				ensure!(path.get(0).map_or(false, |a| a == &credit_asset), Error::<T>::InvalidPath);
+				ensure!(amount_in_max > Zero::zero(), Error::<T>::ZeroAmount);
+				ensure!(amount_out > Zero::zero(), Error::<T>::ZeroAmount);
+
+				Self::validate_swap_path(&path)?;
+				let path = Self::balance_path_from_amount_out(amount_out, path)?;
+
+				let amount_in = path.first().map(|(_, a)| *a).ok_or(Error::<T>::InvalidPath)?;
+				ensure!(
+					amount_in <= amount_in_max,
+					Error::<T>::ProvidedMaximumNotSufficientForSwap
+				);
+
+				Ok((path, amount_in))
+			};
+			let (path, amount_in) = match inspect_path(credit_in.asset()) {
+				Ok((p, a)) => (p, a),
+				Err(e) => return Err((credit_in, e)),
+			};
+
+			let (credit_in, credit_change) = credit_in.split(amount_in)?;
+			let credit_out = Self::credit_swap(credit_in, &path)?;
+
+			Self::deposit_event(Event::CreditSwapExecuted { amount_in, amount_out, path });
+
+			Ok((credit_out, credit_change))
```

Comment:
This is different than `swap_tokens_for_exact_tokens`, which would provide only the amount consumed. I understand the need to return `credit_change` here, but why `credit_out`? Doesn't success imply the `amount_out` that it was called with?

Reply:
I think I see, the `amount_out` is a `Balance`, but you actually need to create the `Credit`?

Reply:
yeah, we get the `credit_in`, and give back the `credit_out`. no `to` or `from` addresses